### ISSUE TITLE
Should the post-ksh93u support for AST magic paths be retained or removed?

### ISSUE THREAD

**Issue by krader1961** [att/ast#510]
A new feature was introduced after the ksh93u+ release that is, AFAICT, only documented in the source (and cryptically at that). Primarily file *src/lib/libast/path/pathcanon.c*. This new code is the cause of regression issue #492. Below are some of the relevant source code comments. The question is whether we want to retain this new code as there is at least one significant bug, the `pathdev()` function is too long and hard to understand, and the feature is poorly documented. Issue #492 would be most easily fixed by simply reinstating the older `pathcanon()` implementation.

```
 * check for ast pseudo dev/attribute paths and optionally canonicalize
```

```
 * path name canonicalization -- preserves the logical view
 *
 *      remove redundant .'s and /'s
 *      move ..'s to the front
 *      /.. preserved (for pdu and newcastle hacks)
 *      if (flags&PATH_ABSOLUTE) then pwd prepended to relative paths
 *      if (flags&PATH_PHYSICAL) then symlinks resolved at each component
 *      if (flags&(PATH_DOTDOT|PATH_PHYSICAL)) then each .. checked for access
 *      if (flags&PATH_EXISTS) then path must exist at each component
 *      if (flags&PATH_VERIFIED(n)) then first n chars of path exist
 *
 * longer pathname possible if (flags&PATH_PHYSICAL) involved
 * 0 returned on error and if (flags&(PATH_DOTDOT|PATH_EXISTS)) then canon
 * will contain the components following the failure point
 *
 * pathcanon() and pathdev() return pointer to trailing 0 in canon
 * pathdev() handles ast specific /dev/ and /proc/ special files
 * pathdev(PATH_DEV) returns 0 if path is not a valid special file
 * pathdev(PATH_AUX) Pathdev_t.flags|=PATH_AUX if AUX fd was created
 * see pathopen() for the api that ties it all together
```

```
 * NAMED_XATTR
 *
 *      0:disabled
 *      1:enabled if O_XATTR!=0
 *
 * /dev/file/xattr@relative-path/file//@//?(attr)
 * /dev/file/xattr@/absolute-path/file//@//?(attr)
 *
 * names the xattr file "attr" for "file" or the xattr dir for "file"
 * neither "relative-path/file" nor "absolute-path/file" may contain "//@//"
 * so pathcanon() "relative-path/file" and "absolute-path/file" before
 * constructing a /dev/file/xattr@/...//@// path
```

**Comment by krader1961** [comment]
Note that the named xattr support is apparently considered a work in progress by the AST team because it is still protected by `#if NAMED_XATTR` directives. So we could easily take a baby step towards simplifying the code by simply removing all the code predicated on that preprocessor symbol.

**Comment by krader1961** [comment]
Also, if we're going to retain this new feature then the conditional compilation of it (e.g., `#if NAMED_XATTR`) should be converted to unconditional compilation. Such as we've been doing with the `SHOPT_`&ast; symbols. Code like this should never be committed if it isn't ready for prime-time. At my $dayjob we only allow such conditional compilation for a short, well defined, transition interval. I am obviously not a fan of the AST team's approach to implementing new features that leaves the code with lots of `#if new-feature...#endif` blocks many years after the team obviously considered the feature no longer new or optional.

**Comment by siteshwar** [comment]
I think removing this feature is going to bring back [this bug](https://github.com/att/ast/issues/462#issuecomment-381270571). But at the same time, I am afraid retaining it is going to create more problems than it solves. A shell should not see file system differently than external utilities it forks. I vote to drop this feature.

**Comment by DavidMorano** [comment]
For those who do not know or did not already guess,
I think this so called "feature" of overlay path names came from the 3DFS (three dimensional file system) code. I cannot be totally sure, but I think it quite likely. That code (3DFS) used to be regularly shipped w/ KSH but in some relatively recent release it was no longer shipped (at least not in the form that it had been before).  The code in the old 3DFS used to provide these sorts of path overlays. I think the code in |pathdev()| was an attempt to being some of the old path-overlay features directly into KSH without having to have the 3DFS system loaded. In the old AT&T AST environment, they used to run w/ the 3DFS system loaded almost all of the time.  Their main build option used to depend on this 3DFS feature set (they had other build options which did not depend on it though). Anyway, my guess is that there were at least some customers who wanted the old 3DFS path overlays even though the 3DFS system itself was not loaded. Why the 3DFS system was no longer shipped as before (and therefore available for use by users who wanted it) I do not know. Someone from AT&T AST (if there is anyone left now) probably knows why this shift occurred with the status of 3DFS.  Maybe support for the whole of 3DFS was failing, so it might have been thought to bring the path-overlay feature into KSH directly before 3DFS went away completely. But again, I am really only guessing on this stuff.

**Comment by DavidMorano** [comment]
Yes, the code in |pathdev()| is -- in my opinion -- poorly written for readability and maintainability (the primary goals, only behind proper function) for modern code. It looks like whoever wrote it (probably one of the original 3DFS guys from decades ago, before being adopted into KSH proper) seems to have written it for supposed speed reasons. But again, in my opinion, speed is not as important as readability and maintainability now-a-days. And the supposed speed advantage of parsing out path names by comparing characters one at a time in this tree-like manner is not justified by the mess it creates in the source code.

Thanks to @krader1961 for fixing the bug that was in there. I do not know if I would have had your patience to go through the code as you did. I hate to say it, but if the code is eventually retained (probably only because some customer had requested it due to the loss of the real 3DFS system) it is a very good candidate to be refactored, to be more readable and obvious about what it is doing.

**Comment by kernigh** [comment]
> Also, if we're going to retain this new feature then the conditional compilation of it (e.g., #if NAMED_XATTR) should be converted to unconditional compilation.

NAMED_XATTR uses the O_XATTR flag to open(2). This is part of [fsattr(5) in Illumos](https://illumos.org/man/5/fsattr) (or Solaris), and probably doesn't exist in other systems. The code wouldn't work with [xattr(7) in Linux](http://man7.org/linux/man-pages/man7/xattr.7.html) nor [extattr(2) in FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=extattr&apropos=0&sektion=2&manpath=FreeBSD+11.1-RELEASE+and+Ports&arch=default&format=html). The current version of pathcanon.c [enables NAMED_XATTR if the system has O_XATTR](https://github.com/att/ast/blob/d952fdfb4630ded4cb1f5387f527bc1b24e89f3c/src/lib/libast/path/pathcanon.c#L85).

> A shell should not see file system differently than external utilities it forks.

If more programs would use libast, then magic paths might be useful. I only have ksh using libast, so magic paths seem useless.

Also, magic paths don't always work in ksh:

```
$ echo line >/tmp/line
$ exec 3</tmp
$ exec 4</proc/self/fd/3/line
$ cat </proc/self/fd/4
/home/kernigh/park/ast/build/src/cmd/ksh93/ksh: /proc/self/fd/4: cannot open [No such file or directory]
$ cat <&4
line
```

`/proc/self/fd/3/line` works but `/proc/self/fd/4` doesn't work. This is in OpenBSD with no /proc filesystem, so /proc can only work as a magic path.

**Comment by krader1961** [comment]
I'm going to resolve this issue by removing the `pathdev()` function and reinstating the ksh93u+ implementation of `pathcanon()` this weekend if no one makes any objections.

**Comment by DavidMorano** [comment]
@krader1961 
> I'm going to resolve this issue by removing the pathdev() function and reinstating the ksh93u+ implementation of pathcanon() this weekend if no one makes any objections.

Note that BASH also supports (for many many years now) some magic paths, specifically:
/dev/tcp/host/port
and
/dev/udp/host/port

where 'host' and 'port' are arguments for a host name or address, et cetera.

These above magic paths are implemented in KSH with the |pathdecv()| subroutine and perhaps nowhere else, so one consequence of eliminating |pathdev()| might be to break this particular compatibility with BASH. How these magic paths (above) were implemented previously, I am not sure (maybe just w/ 3DFS). Whether some people actually complain about the lack of these magic paths or not, I have no idea.  I guess it depends on how much these magic paths have been used out in the wild. I have used these (magic paths) myself on occasion, but not for anything portable or very important (I have many other magic paths which do both similar and also exactly the same function as these magic paths above).  Anyway, just something to keep in mind if any future complaints arise.

**Comment by krader1961** [comment]
@DavidMorano It is not at all obvious the `pathdev()` function is necessary to support paths like */dev/tcp/host/$port*. Did those paths work in the ksh93u release (which did not have that function)?

**Comment by DavidMorano** [comment]
@krader1961 
> Did those paths work in the ksh93u release (which did not have that function)?

Good question!

I honesty do not remember which release did not have them and when they became native to KSH.
There was a time, some years ago (not really sure), when KSH did not have these (at least not /dev/tcp/$host/$port).  Rather these paths were only (as far as I remember) supported with the use of the 3DFS system (three dimensional file-system). But at some point, these paths became native to KSH, but I do not remember when.  It could have only become native with the 93v- release, but I've been on that branch a fairly long time, even before it was made official and loaded to GitHub (I originally switched to that branch from when it was beta on the old ATT/AST research site).

Also, I have not even investigated the current code to know exactly where those paths are supported.
Yes, |pathdev()| certainly checks for those paths, but I suppose that the paths could be (or also be) supported elsewhere.

In Lieu of doing a full and careful investigation of the code, one could do the actual experiment, namely:

1) verify as a control that something like 
  $ cat < /dev/tcp/localhost/daytime

is working on the present KSH. Of course, make sure that a TCP server is available and provides some output on whatever port is used.

2) Take |pathdev()| out of KSH and recompile

3) On the new KSH (without the old |pathdev()|) see if
  $ cat < /dev/tcp/localhost/daytime

still works.

In a related matter, there is the question: Should KSH continue to support the /dev/tcp/... and /dev/udp/... paths? And my answer would be YES. But my answer is bolstered by the fact that these paths have been supported in BASH for what seems like over ten or more years (maybe longer).  There have been several advertisements or other articles over at least the last ten years on how to creatively use these paths from within BASH.

**Comment by krader1961** [comment]
FWIW, ksh93u supports magic paths like */dev/tcp/$hostname/$port* but does so via the `sh_pathopen()` function. In ksh93u the `pathdev()` function does not exist. Support for those magic paths was moved into the `pathdev()` function in the ksh93v beta release. I've already reinstated the ksh93u implementation of `sh_pathopen()` so the `master` branch no longer depends on `pathdev()` to support those magic paths in the `ksh` program. Subsequent changes will remove the `pathopen()` and `pathdev()` functions.
